### PR TITLE
WIP Rx state/connect partial obs and signals

### PR DESCRIPTION
**Pull request by mikelgo**
WIP PR for #1632 

Tests are still in WIP

### CODE REVIEW

File: apps/docs/docs/state/tutorials/basic-tutorial/01-setup.md [hunk 1]
```diff
@@ -87,7 +87,7 @@ We will use `set()` as we already have initial values assigned to the `initCompo
 
 ```typescript
  constructor(...) {
-  ...
+...
```
Comment (mikelgo) [inline apps/docs/docs/state/tutorials/basic-tutorial/01-setup.md:90]: revert. Don't know where this is comming from

**Comment by mikelgo** [comment]
@hoebbelsB as we had the initial discussion you might want to check this PR.
Tests are still WIP (need to get familiar with the `scheduled` helper and how to use it with signals correctly)

File: libs/state/src/lib/convert-partials-to.observable.ts [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+  return combineLatest(observables).pipe(
+    map((values) => {
+      const result: Partial<T> = {} as Partial<T>;
+      for (let i = 0; i < keys.length; i++) {
+        // @ts-ignore
+        result[keys[i]] = values[i];
+      }
+      return result;
+    }),
+    coalesceWith(timer(0))
```
Comment (mikelgo) [inline libs/state/src/lib/convert-partials-to.observable.ts:43]: is this the correct coalescing operator?

File: libs/state/src/lib/convert-partials-to.observable.ts [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+
+    if (isObservable(obsOrSignal)) {
+      return obsOrSignal?.pipe(startWith(undefined));
+    }
+    if (isSignal(obsOrSignal)) {
+      return toObservable(obsOrSignal, { injector });
+    }
+  });
+
+  return combineLatest(observables).pipe(
```
Comment (mikelgo) [inline libs/state/src/lib/convert-partials-to.observable.ts:34]: should I filter out any key which has a value of undefined?
Undefined might come from line 27

File: libs/state/src/lib/convert-partials-to.observable.ts (hunk 2 above)
Comment (hoebbelsB) [reply libs/state/src/lib/convert-partials-to.observable.ts:43]: it is the correct operator, but why would you coalesce it on the next `macrotask`? It makes much more sense to coalesce on `microtask` instead. Also, it shouldn't trigger ngZone.

Please try with `coalesceWith(from(Promise.resolve()))` and use the unpatched promise from the rx angular cdk.

We have to write tests though, that confirm `RxState` is not running out of zone. That could potentially cause issues when state selections are running out of zone.

Comment (hoebbelsB) [reply libs/state/src/lib/convert-partials-to.observable.ts:43]: also, I think coalescing should happen before the mapping. This way, the mapping would be less often executed

File: libs/state/src/lib/rx-state.service.ts [hunk 4]
```diff
@@ -394,32 +396,53 @@ export class RxState<T extends object> implements OnDestroy, Subscribable<T> {
...
       | Observable<T[K] | V>
       | Signal<T[K] | V>,
     projectValueFn?: ProjectValueReducer<T, K, V>
   ): void {
+    if (
+      isPartialOfSignalsOrObservablesGuard(keyOrInputOrSliceOrPartials$) &&
+      !projectOrSlices$ &&
+      !projectValueFn
+    ) {
+      console.log('isPartialOfSignalsOrObservablesGuard');
```
Comment (hoebbelsB) [inline libs/state/src/lib/rx-state.service.ts:424]: should be removed :)

File: libs/state/src/lib/convert-partials-to.observable.ts (hunk 2 above)
Comment (mikelgo) [reply libs/state/src/lib/convert-partials-to.observable.ts:43]: makes absolutely sense, will rework it soon

File: libs/state/src/lib/convert-partials-to.observable.ts [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+    coalesceWith(from(Promise.resolve())),
+    map((values) => {
+      const result: Partial<T> = {} as Partial<T>;
+      for (let i = 0; i < keys.length; i++) {
+        // @ts-ignore
+        result[keys[i]] = values[i];
```
Comment (edbzn) [inline libs/state/src/lib/convert-partials-to.observable.ts:35]: Can we get rid of the comment here?

File: libs/state/src/lib/convert-partials-to.observable.ts [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+export function convertPartialsToObservable<
+  T extends object,
+  K extends keyof T
+>(
+  input: Partial<{ [K: string]: Observable<T[K]> | Signal<T[K]> }>,
+  injector?: Injector
+): Observable<Partial<T>> {
+  const keys = Object.keys(input);
+
+  // eslint-disable-next-line @typescript-eslint/no-unsafe-return
```
Comment (edbzn) [inline libs/state/src/lib/convert-partials-to.observable.ts:17]: I would get rid of this comment and probably throw like so:

 ```ts
 const observables = keys.map((key) => {
    const obsOrSignal = input[key];

    if (isObservable(obsOrSignal)) {
      return obsOrSignal?.pipe(startWith(undefined));
    }
    if (isSignal(obsOrSignal)) {
      return toObservable(obsOrSignal, { injector });
    }

    throw new TypeError(`Expected type Observable or Signal for key "${key}"`);
  });
  ```
  
  Wdyt?

File: libs/state/selections/src/lib/utils/guards.ts [hunk 7]
```diff
@@ -69,3 +70,22 @@ export function isStringArrayFunctionAndOptionalObjectTupleGuard<R>(
...
+export function isPartialOfSignalsOrObservablesGuard<T, K extends keyof T>(
+  arg: unknown
+): arg is Partial<{ [K: string]: Observable<T[K]> | Signal<T[K]> }> {
+  if (!isObjectGuard(arg)) {
+    return false;
+  }
+  for (const key in arg) {
+    // eslint-disable-next-line no-prototype-builtins
+    if (arg.hasOwnProperty(key)) {
+      // @ts-ignore
```
Comment (edbzn) [inline libs/state/selections/src/lib/utils/guards.ts:83]: I would also get rid of this comment.

File: libs/state/src/lib/rx-state.service.ts [hunk 8]
```diff
@@ -394,32 +396,51 @@ export class RxState<T extends object> implements OnDestroy, Subscribable<T> {
...
+        inputOrSlice$ = keyOrInputOrSliceOrPartials$;
       } else {
         // why can't typescript infer the correct type?
         inputOrSlice$ = toObservable(
-          keyOrInputOrSlice$ as Signal<Partial<T> | V>,
+          keyOrInputOrSliceOrPartials$ as Signal<Partial<T> | V>,
           { injector: this.injector }
         );
       }
     }
```
Comment (edbzn) [inline libs/state/src/lib/rx-state.service.ts:440]: I think this nested if/else branching starts to be hard to read. I think there's a better way to handle overloads, would be nice to refactor it at some point.

File: libs/state/src/lib/convert-partials-to.observable.ts (hunk 6 above)
Comment (mikelgo) [reply libs/state/src/lib/convert-partials-to.observable.ts:17]: Like!

File: libs/state/src/lib/rx-state.service.ts (hunk 8 above)
Comment (mikelgo) [reply libs/state/src/lib/rx-state.service.ts:440]: Totally agree. Overloads in Typescript are a complete mess

File: libs/state/selections/src/lib/utils/guards.ts (hunk 7 above)
Comment (mikelgo) [reply libs/state/selections/src/lib/utils/guards.ts:83]: done

File: libs/state/src/lib/convert-partials-to.observable.ts (hunk 6 above)
Comment (mikelgo) [reply libs/state/src/lib/convert-partials-to.observable.ts]: done

File: libs/state/src/lib/convert-partials-to.observable.ts (hunk 5 above)
Comment (mikelgo) [reply libs/state/src/lib/convert-partials-to.observable.ts]: done

File: libs/state/src/lib/rx-state.service.ts (hunk 8 above)
Comment (mikelgo) [reply libs/state/src/lib/rx-state.service.ts:440]: @edbzn I started in a separate PR to simplify the `connect`-method: https://github.com/rx-angular/rx-angular/pull/1652 
wdyt? 
To me it looks better. However it is not finished yet. There are two failing test cases bc my guards are not perfect yet I believe

**Comment by eneajaho** [comment]
Hello @mikelgo anything I can help with? In order to move this PR forward?

**Comment by mikelgo** [comment]
> Hello @mikelgo anything I can help with? In order to move this PR forward?

hey @eneajaho thanks for asking. 
I just lack a bit of time right now :D And I'm struggling a bit on how to write the tests in the same way as the existing tests using the `scheduled`-helper from `rxjs`. I would need to do some conversion from signals to observable to use this helper, but not sure if this would be really the way to go (I would not do it).

**Pull request closed without merging**